internal/overlord/logstate/forwarder.go

```diff
@@ -0,0 +1,226 @@
+}
+
+// newLogForwarderInternal is shared by the real and test code.
+func newLogForwarderInternal(
+	service string, target *plan.LogTarget, iterator servicelog.Iterator,
+	client logClient, maxMessages int, delay time.Duration,
+) *logForwarder {
+
+	return &logForwarder{
+		service:          service,
+		target:           target,
+		iterator:         iterator,
+		maxMessages:      maxMessages,
+		delay:            delay,
+		client:           client,
+		stopCh:           make(chan struct{}),
+		doneWithIterator: make(chan struct{}),
+	}
+}
+
+// forward continually receives logs from the iterator and buffers them. It
+// flushes the buffer after a timeout, or when a certain number of messages
+// have been written.
+// This method will block until the service is stopped, or stop() is called.
+// Hence, it should be run in a separate goroutine.
+func (f *logForwarder) forward() {
+	// Create a timer and immediately stop it, so it starts unset
+	flushTimer := time.NewTimer(0)
+	stopTimer(flushTimer)
+
+	cancelIterator := make(chan struct{})
+	messagesCh := make(chan servicelog.Entry)
+	go f.iterate(cancelIterator, messagesCh)
+
+	var buffer []servicelog.Entry
+	for {
+		select {
+		case message := <-messagesCh:
+			buffer = append(buffer, message)
+
+			if len(buffer) >= f.maxMessages {
+				// Buffer full - send logs to target
+				stopTimer(flushTimer)
+				f.sendLogs(buffer)
+				buffer = buffer[:0]
+
+			} else if len(buffer) == 1 {
+				// First write - set the timeout
+				stopTimer(flushTimer)
+				flushTimer.Reset(f.delay)
+			}
+
+		case <-flushTimer.C:
+			f.sendLogs(buffer)
+			buffer = buffer[:0]
+
+		case <-f.stopCh:
+			// Cancel the iterator channel, so it will stop waiting for more data
+			close(cancelIterator)
+
+			// Drain remaining messages from parser
+			for message := range messagesCh {
+				buffer = append(buffer, message)
+				if len(buffer) >= f.maxMessages {
+					f.sendLogs(buffer)
+					buffer = buffer[:0]
+				}
+			}
+
+			// Once the above for loop returns, the iterate method has finished,
+			// so we are done using the iterator.
+			close(f.doneWithIterator)
+			f.sendLogs(buffer)
+
+			err := f.client.Close()
+			if err != nil {
+				logger.Noticef("Internal error: closing log client %q -> %q: %s",
+					f.service, f.target.Name, err)
+			}
+			return
+		}
+	}
+}
+
+// iterate continually pulls logs from the iterator/parser, and sends these to
+// the forward loop via the messages channel. When the provided cancel channel
+// is closed, iterate will pull all remaining logs from the parser before
+// exiting.
+func (f *logForwarder) iterate(cancel <-chan struct{}, messages chan servicelog.Entry) {
+	parser := servicelog.NewParser(f.iterator, parserSize)
+
+	for f.iterator.Next(cancel) {
+		for parser.Next() {
+			// The forward loop is either selecting on this channel or draining it
+			// in the cancel case. So no need for a select here.
+			messages <- parser.Entry()
```

Comment:
The machinery that decouples the iterator, the parser, the buffer inside the forward loop, and the client, seem to add signifciant complexity that we don't seem to benefit from. Let's just take a step back and consider the overall flow.

At the moment we have data buffered at those layers:

- The ring buffer
- (NOT the iterator, it's just a pointer to the ring)
- The servicelog.Parser
- The buffer inside forward()
- The logClient (whether persistent or not, there's a data buffer there while sending)

All of those intermediate layers are just processing data from one side to the other. It might be tempting to buffer data in between to make sure the system does not run out of memory, well, except both sides are _already in memory!_  So any attempt to buffer data in the middle in an intermediate state is actually _increasing_ the amount of memory used in the overall system. If we want less memory consumption, all we need to do is to consume more slowly from the left side, and when we do consume, send it all the way to the right side.

Assuming such a design, we'd need a single goroutine that parses out of the iterator, and any time it manages to parse anything it adds to the client which can then serialize to its buffer and keep it pretty much ready for dispatch. Straight all the way from ring buffer to the output buffer.

Then, there's apparently another issue in the current system: there is an independent client for each forwarder, which means if we have 10 services and 2 targets, we have 20 different connections, each with their own series of buffers. Is that the case?  If so, that really doesn't feel so great.

So, the recommended simplified design would be something like this: One single client per log target for all of Pebble, no intermediate buffers.  All the data that leaves the ring buffer is processed on one single control goroutine, and lands on the client buffer as ready to dispatch as possible, so we know the actual amount of bytes we have to send, and are not holding references to Go structures. One _Push_ blocking method to add entries into the client which also flushes when above a given size limit, and a second method _Flush_ to force it empty.  These methdos make it easy for the client to tolerate concurrency properly (any number of goroutines pushing their messages will naturally block and wait when a size limit is exceeded. To handle the time-based delay, we can have a second method like _FlushAfter(duration),_ where the timing of last flush is held internally by the client, which means concurrency would also work fine if multilpe goroutines are recommending the flush at once.

And that would be all, I think.  Of course, it's a bit hard to keep the whole picture at once in memory, but hopefully I'm not missing major points here.